src/v/kafka/server/group_metadata_migration.cc

```diff
@@ -0,0 +1,632 @@
+      .then([](std::vector<cluster::topic_result> res) {
+          /*
+           * kindly ask client to retry on error
+           */
+          vassert(res.size() == 1, "expected exactly one result");
+          if (res[0].ec != cluster::errc::success) {
+              vlog(
+                mlog.warn,
+                "can not create __consumer_offsets topic - {}",
+                res[0].ec);
+              return false;
+          }
+          return true;
+      })
+      .then([&controller, timeout](bool success) {
+          if (success) {
+              return controller.get_api()
+                .local()
+                .wait_for_topic(model::kafka_consumer_offsets_tn, timeout)
+                .then([](std::error_code ec) { return !ec; });
+          }
+          return ss::make_ready_future<bool>(success);
+      })
+      .handle_exception([](const std::exception_ptr& e) {
+          vlog(mlog.warn, "can not create __consumer_offsets topic - {}", e);
+          // various errors may returned such as a timeout, or if the
+          // controller group doesn't have a leader. client will retry.
+          return false;
+      });
+}
+
+ss::future<group_recovery_consumer_state> recover_group_state(
+  model::ntp ntp,
+  ss::shard_id shard,
+  ss::sharded<cluster::partition_manager>& pm,
+  ss::sharded<ss::abort_source>& as,
+  group_metadata_serializer_factory serializer_f) {
+    using ret_t = group_recovery_consumer_state;
+    return pm.invoke_on(
+      shard,
+      [ntp = std::move(ntp), &as, serializer_f = std::move(serializer_f)](
+        cluster::partition_manager& pm) mutable {
+          auto p = pm.get(ntp);
+          if (!p) {
+              return ss::make_ready_future<ret_t>(ret_t{});
+          }
+          auto reader_cfg = storage::log_reader_config(
+            p->start_offset(),
+            model::offset::max(),
+            ss::default_priority_class());
+
+          return p->make_reader(reader_cfg)
+            .then([&as = as.local(), serializer_f = std::move(serializer_f)](
+                    model::record_batch_reader reader) mutable {
+                return std::move(reader).consume(
+                  group_recovery_consumer(serializer_f(), as),
+                  model::no_timeout);
+            });
+      });
+}
+
+bool are_offsets_equal(
+  const group::offset_metadata& lhs, const group::offset_metadata& rhs) {
+    return lhs.metadata == rhs.metadata && lhs.offset == rhs.offset
+           && lhs.committed_leader_epoch == rhs.committed_leader_epoch;
+}
+bool are_stms_equivalent(const group_stm& source, const group_stm& target) {
+    if (source.get_metadata() != target.get_metadata()) {
+        vlog(
+          mlog.debug,
+          "source and target group metadata does not match, source: {}, "
+          "target: {}",
+          source.get_metadata(),
+          target.get_metadata());
+        return false;
+    }
+
+    for (auto& [tp, o] : source.offsets()) {
```

Comment:
Is it worth doing a size check just prior to this? I'm not sure how extra `tp`s would get in target, but still.